mmrazor/models/fake_quants/lsq.py

```diff
@@ -0,0 +1,137 @@
+        # to make it learnable.
+        if self.observer_enabled[0] == 1:
+            self.activation_post_process(X.detach())
+            _scale, _zero_point = \
+                self.activation_post_process.calculate_qparams()
+            _scale = _scale.to(self.scale.device)
+            _zero_point = _zero_point.to(self.zero_point.device)
+
+            if self.ch_axis != -1:
+                self.scale.data = torch.ones_like(_scale)
+                self.zero_point.data = torch.zeros_like(_zero_point.float())
+
+            self.scale.data.copy_(_scale)
+            self.zero_point.data.copy_(_zero_point.float())
+        else:
+            self.scale.data.abs_()
+            self.scale.data.clamp_(min=self.eps.item())
+
+        if self.fake_quant_enabled[0] == 1:
+            if _is_symmetric_quant(self.qscheme):
+                self.zero_point.data.zero_()
+            else:
+                self.zero_point.data.clamp_(self.quant_min,
+                                            self.quant_max).float()
+
+            if self.is_per_channel:
+                if self.use_grad_scaling:
+                    grad_factor = 1.0 / (X.numel() / X.shape[self.ch_axis] *
+                                         self.quant_max)**0.5
+                else:
+                    grad_factor = 1.0
+                if is_tracing_state():
+                    X = FakeQuantizeLearnablePerchannelAffine.apply(
+                        X, self.scale, self.zero_point, self.ch_axis,
+                        self.quant_min, self.quant_max, grad_factor)
+                else:
+                    X = _fake_quantize_learnable_per_channel_affine_training(
+                        X, self.scale, self.zero_point, self.ch_axis,
+                        self.quant_min, self.quant_max, grad_factor)
+            else:
+                if self.use_grad_scaling:
+                    grad_factor = 1.0 / (X.numel() * self.quant_max)**0.5
+                else:
+                    grad_factor = 1.0
+                X = torch._fake_quantize_learnable_per_tensor_affine(
+                    X, self.scale, self.zero_point, self.quant_min,
+                    self.quant_max, grad_factor)
+        return X
+
+
+def _fake_quantize_learnable_per_channel_affine_training(
+        x, scale, zero_point, ch_axis, quant_min, quant_max, grad_factor):
+    zero_point = (zero_point.round() - zero_point).detach() + zero_point
+    new_shape = [1] * len(x.shape)
+    new_shape[ch_axis] = x.shape[ch_axis]
+    scale = grad_scale(scale, grad_factor).reshape(new_shape)
+    zero_point = grad_scale(zero_point, grad_factor).reshape(new_shape)
+    x = x / scale + zero_point
+    x = (x.round() - x).detach() + x
+    x = torch.clamp(x, quant_min, quant_max)
+    return (x - zero_point) * scale
+
+
+def grad_scale(t, scale):
+    return (t - (t * scale)).detach() + (t * scale)
+
+
+class FakeQuantizeLearnablePerchannelAffine(torch.autograd.Function):
```

Comment:
need to take deployment into consideration later

Reply:
Then update when need.